File: libs-logwriter/src/main/java/com/rzrasel/logwriter/backup/LogWriter.java
package com.rzrasel.logwriter.backup;

import android.annotation.SuppressLint;

public class LogWriter {
    protected static String FULL_CLASS_NAME = LogWriter.class.getName();
    private static final String CLASS_NAME = LogWriter.class.getSimpleName();
    private static final String DEBUG_PREFIX = "DEBUG_LOG_WRITER_PRINT: ";
    //private static final String LOG_FORMAT = "Message: %s | Class: %s.%s | Method: %s | Line: %d";
    private static final String LOG_FORMAT = "Message: %s - Class: %s - Method: %s - Line Number: %d | Full Class | Package: %s | Class: %s | Line Number: %d";
    private static final String RESET = LogType.RESET.getColorCode();

    public LogWriter() {
        debugPrint("Initialized LogWriterV2 for class: " + getCallerClassName());
    }

    public static void Log(String message, boolean isDebug) {
        log(message, isDebug);
    }

    public static void Log(String message) {
        log(message);
    }

    public static void log(String message, boolean isDebug) {
        if(isDebug) {
            log(message);
        }
    }

    public static void log(String message) {
        /*if(!BuildConfig.DEBUG) {
            return;
        }*/
        StackTraceElement callerStackTraceElement = getCallerStackTraceElement();
        if(callerStackTraceElement == null) {
            debugPrint("[Unknown Caller] Unable to determine caller information for message: " + message);
            return;
        }

        String packageName = getCallerPackageName(callerStackTraceElement);
        packageName = packageName.isEmpty() ? "(default package)" : packageName;
        String className = callerStackTraceElement.getClassName();
        String simpleClassName = className.substring(className.lastIndexOf(".") + 1);
        String methodName = callerStackTraceElement.getMethodName();
        int lineNumber = callerStackTraceElement.getLineNumber();

        @SuppressLint("DefaultLocale")
        String logMessage = String.format(
                LOG_FORMAT,
                message,
                simpleClassName,
                methodName,
                lineNumber,
                packageName + ".",
                className,
                lineNumber
        );

        debugPrint(logMessage);
    }

    private static StackTraceElement getCallerStackTraceElement() {
        StackTraceElement[] stackTrace = Thread.currentThread().getStackTrace();
        boolean isClassFound = false;
        String targetClassName = FULL_CLASS_NAME;
        for(StackTraceElement stackTraceElement : stackTrace) {
            if(stackTraceElement.getClassName().equalsIgnoreCase(targetClassName)) {
                isClassFound = true;
            }
            if(isClassFound && !stackTraceElement.getClassName().equalsIgnoreCase(targetClassName)) {
                return stackTraceElement;
            }
        }
        return null;
    }

    private static String getCallerPackageName(StackTraceElement callerElement) {
        String className = callerElement.getClassName();
        if(className == null || className.isEmpty()) {
            return "";
        }
        int lastDot = className.lastIndexOf('.');
        return lastDot > 0 ? className.substring(0, lastDot) : "";
    }

    private static String getCallerClassName() {
        StackTraceElement callerElement = getCallerStackTraceElement();
        return callerElement != null ? callerElement.getClassName() : "Unknown";
    }

    private static String getCallerMethodName() {
        StackTraceElement callerElement = getCallerStackTraceElement();
        return callerElement != null ? callerElement.getMethodName() : "Unknown";
    }

    private static int getCallerLineNumber() {
        StackTraceElement callerElement = getCallerStackTraceElement();
        return callerElement != null ? callerElement.getLineNumber() : -1;
    }

    private static void debugPrint(String message) {
        String coloredMessage = DEBUG_PREFIX + message;
        System.out.println(coloredMessage);
    }
}

File: libs-logwriter/src/main/java/com/rzrasel/logwriter/LogWriter.kt
package com.rzrasel.logwriter

import android.annotation.SuppressLint
import android.content.Context

object LogWriter {
    private val FULL_CLASS_NAME = LogWriter::class.java.name
    private val CLASS_NAME = LogWriter::class.java.simpleName
    private const val DEBUG_PREFIX = "DEBUG_LOG_WRITER_PRINT: "
    private const val LOG_FORMAT = "Message: %s - Class: %s - Method: %s - Line Number: %d | Full Class | Package: %s | Class: %s | Line Number: %d"

    init {
        debugPrint("Initialized LogWriter for class: ${getCallerClassName()}")
    }

    @JvmStatic
    fun log(message: String?) = log(message, context = null, isDebug = true)

    @JvmStatic
    fun log(message: String?, context: Context? = null, isDebug: Boolean = true) {
        if (context != null && !BuildConfig.isDebuggable(context)) return
        if (!isDebug) return

        val callerStackTraceElement = getCallerStackTraceElement()
        if (callerStackTraceElement == null) {
            debugPrint("[Unknown Caller] Unable to determine caller information for message: $message")
            return
        }

        val packageName = getCallerPackageName(callerStackTraceElement)
        val actualPackageName = packageName.ifEmpty { "(default package)" }
        val className = callerStackTraceElement.className
        val simpleClassName = className.substringAfterLast(".")
        val methodName = callerStackTraceElement.methodName
        val lineNumber = callerStackTraceElement.lineNumber

        @SuppressLint("DefaultLocale")
        val logMessage = String.format(
            LOG_FORMAT,
            message,
            simpleClassName,
            methodName,
            lineNumber,
            "$actualPackageName.",
            className,
            lineNumber
        )

        debugPrint(logMessage)
    }

    private fun getCallerStackTraceElement(): StackTraceElement? {
        val stackTrace = Thread.currentThread().stackTrace
        var isClassFound = false
        val targetClassName = FULL_CLASS_NAME

        for (stackTraceElement in stackTrace) {
            if (stackTraceElement.className.equals(targetClassName, ignoreCase = true)) {
                isClassFound = true
            }
            if (isClassFound && !stackTraceElement.className.equals(targetClassName, ignoreCase = true)) {
                return stackTraceElement
            }
        }
        return null
    }

    private fun getCallerPackageName(callerElement: StackTraceElement): String {
        val className = callerElement.className
        if (className.isNullOrEmpty()) {
            return ""
        }
        val lastDot = className.lastIndexOf('.')
        return if (lastDot > 0) className.take(lastDot) else ""
    }

    private fun getCallerClassName(): String {
        val callerElement = getCallerStackTraceElement()
        return callerElement?.className ?: "Unknown"
    }

    private fun getCallerMethodName(): String {
        val callerElement = getCallerStackTraceElement()
        return callerElement?.methodName ?: "Unknown"
    }

    private fun getCallerLineNumber(): Int {
        val callerElement = getCallerStackTraceElement()
        return callerElement?.lineNumber ?: -1
    }

    private fun debugPrint(message: String) {
        val coloredMessage = DEBUG_PREFIX + message
        println(coloredMessage)
    }
}